### ISSUE TITLE
dlclose() causes .so unloaded and does not reload it back on reverse-continue

### ISSUE THREAD

**Issue by midenok** [rr-debugger/rr#3846]
In continuation of #3845.

For the trace of the following sequence:
1. dlopen(X.so)
2. debugged code
3. dlclose(X.so)
4. After dlclose()

Until 4. is not reached 2. displays X.so in `info shared`. But if 4. is reached in trace replay, reverse-continue to 2. does not return back X.so to loaded modules.

The workaround to this is restart from the start or replay dlopen() again. Saved checkpoints also restore correct modules list.

**Comment by khuey** [comment]
@rocallahan this is a known gdb issue isn't it?

**Comment by rocallahan** [comment]
Yes it is. When GDB observes the dlclose call it doesn't consider whethr execution is going forward or backward. @midenok it would be great if you can file a GDB bug about this.

**Comment by GitMensch** [comment]
The GDB bug report seems to be missing: https://sourceware.org/bugzilla/buglist.cgi?product=gdb&quicksearch=dlclose

Note: there's also a related bug to dlopen, also not reported yet: https://sourceware.org/bugzilla/buglist.cgi?product=gdb&quicksearch=dlopen, recognizable if you reverse-step over that, then step-over it again; those _may_ be reported as a single bug.
